core/unit_test/TestMathematicalFunctions.hpp

```diff
@@ -511,6 +554,53 @@ void do_test_math_unary_function(const Arg (&x)[N]) {
 #define TEST_MATH_FUNCTION(FUNC) \
   do_test_math_unary_function<TEST_EXECSPACE, MathUnaryFunction_##FUNC>
 
+#if defined(KOKKOS_COMPILER_CLANG) || defined(KOKKOS_COMPILER_APPLECC) || \
+    defined(KOKKOS_COMPILER_INTEL_LLVM)
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wgnu-zero-variadic-macro-arguments"
+#endif
+
+// Use similar approach as linux syscalls to expand and cast constants
+#define MAKE_TEST_HALF_MATH_FUNCTION(FUNC, T, ...)                             \
+  MAKE_TEST_HALF_MATH_FUNCTION_N(fn, ##__VA_ARGS__, 9, 8, 7, 6, 5, 4, 3, 2, 1, \
+                                 0)                                            \
+  (__VA_ARGS__)
+#define MAKE_TEST_HALF_MATH_FUNCTION_N(FUNC, T, n0, n1, n2, n3, n4, n5, n6, \
+                                       n7, n8, n9, n, ...)                  \
+  TEST_HALF_MATH_FUNCTION##n
+
+#define TEST_HALF_MATH_UNARY_FUNCTION
+
+#if defined(KOKKOS_COMPILER_MSVC)
+#define TEST_HALF_MATH_FUNCTION(...)
+#else
+#define TEST_HALF_MATH_FUNCTION(...) \
+  MAKE_TEST_HALF_MATH_FUNCTION(FUNC, T, ##__VA_ARGS__)
+#define TEST_HALF_MATH_FUNCTION1(FUNC, T, a) \
+  TEST_MATH_FUNCTION(FUNC)({static_cast<T>(a)})
+#define TEST_HALF_MATH_FUNCTION2(FUNC, T, a, b) \
+  TEST_MATH_FUNCTION(FUNC)({static_cast<T>(a), static_cast<T>(b)})
+#define TEST_HALF_MATH_FUNCTION3(FUNC, T, a, b, c) \
+  TEST_MATH_FUNCTION(FUNC)                         \
+  ({static_cast<T>(a), static_cast<T>(b), static_cast<T>(c)})
+#define TEST_HALF_MATH_FUNCTION4(FUNC, T, a, b, c, d) \
+  TEST_MATH_FUNCTION(FUNC)                            \
+  ({static_cast<T>(a), static_cast<T>(b), static_cast<T>(c), static_cast<T>(d)})
+#define TEST_HALF_MATH_FUNCTION5(FUNC, T, a, b, c, d, e)     \
+  TEST_MATH_FUNCTION(FUNC)                                   \
+  ({static_cast<T>(a), static_cast<T>(b), static_cast<T>(c), \
+    static_cast<T>(d), static_cast<T>(e)})
+#define TEST_HALF_MATH_FUNCTION6(FUNC, T, a, b, c, d, e, f)  \
+  TEST_MATH_FUNCTION(FUNC)                                   \
+  ({static_cast<T>(a), static_cast<T>(b), static_cast<T>(c), \
+    static_cast<T>(d), static_cast<T>(e), static_cast<T>(f)})
+#define TEST_HALF_MATH_FUNCTION7(FUNC, T, a, b, c, d, e, f, g) \
+  TEST_MATH_FUNCTION(FUNC)                                     \
+  ({static_cast<T>(a), static_cast<T>(b), static_cast<T>(c),   \
+    static_cast<T>(d), static_cast<T>(e), static_cast<T>(f),   \
+    static_cast<T>(g)})
```

Comment:
Why are we using variadic macros instead of variadic function templates?

Reply:
To match the style of the existing tests  and re-use the code which use `TEST_MATH_FUNCTION` for testing built-in types.

Reply:
I don't understand your answer.  I don't see how that matches the style and why the macro approach is better.

Reply:
The current style uses a macro called `TEST_MATH_FUNCTION`. We cannot use that same function for half types - we need the variadic macro. `TEST_HALF_MATH_FUNCTION` matches this approach.

Reply:
@dalg24: I am getting:
```
kokkos/core/unit_test/TestMathematicalFunctions.hpp:564:51: error: static_cast from 'const float *' to 'float' is not allowed
```

Do you know how to fix it?

Reply:
My bad
```diff
diff --git a/core/unit_test/TestMathematicalFunctions.hpp b/core/unit_test/TestMathematicalFunctions.hpp
index a31232008..84feba7f7 100644
--- a/core/unit_test/TestMathematicalFunctions.hpp
+++ b/core/unit_test/TestMathematicalFunctions.hpp
@@ -560,8 +560,10 @@ void do_test_math_unary_function(const Arg (&x)[N]) {

 template <class Half, class Space, class... Func, class Arg, std::size_t N>
 void do_test_half_math_unary_function(const Arg (&x)[N]) {
+  Half y[N];
+  std::copy(x, x + N, y);  // cast to array of half type
   (void)std::initializer_list<int>{
-      (TestMathUnaryFunction<Space, Func, Arg, N>(static_cast<Half>(x)), 0)...};
+      (TestMathUnaryFunction<Space, Func, Arg, N>(y), 0)...};

   // test if potentially device specific math functions also work on host
   if constexpr (!std::is_same_v<Space, Kokkos::DefaultHostExecutionSpace>)